source/microbit/modaudio.cpp

```diff
@@ -0,0 +1,648 @@
+    if (double_pin) {
+        //Start with output zero; pins 00
+        if (val1 < 0) {
+            timer->CC[0] = FIRST_PHASE_START;
+            // -ve 10
+            timer->CC[2] = FIRST_PHASE_START-val1;
+            // zero 11
+        } else {
+            timer->CC[2] = FIRST_PHASE_START;
+            // +ve 01
+            timer->CC[0] = FIRST_PHASE_START+val1;
+            // zero 11
+        }
+        // Output zero; pins 11.
+        if (val2 < 0) {
+            timer->CC[3] = SECOND_PHASE_START;
+            // -ve 10
+            timer->CC[1] = SECOND_PHASE_START-val2;
+            // zero 00
+        } else {
+            timer->CC[1] = SECOND_PHASE_START;
+            // +ve 01
+            timer->CC[3] = SECOND_PHASE_START+val2;
+            // zero 00
+        }
+        //End with output zero; pins 00
+    } else {
+        timer->CC[0] = FIRST_PHASE_START;
+        timer->CC[1] = FIRST_PHASE_START+val1;
+        timer->CC[2] = SECOND_PHASE_START;
+        timer->CC[3] = SECOND_PHASE_START+val2;
+    }
+    timer->TASKS_CLEAR = 1;
+}
+
+static int32_t audio_ticker(void) {
+    int32_t val1 = previous_value + delta;
+    int32_t next_value = val1 + delta;
+    previous_value = next_value;
+    set_gpiote_output_pulses(val1>>4, next_value>>4);
+    if (sample) {
+        int32_t buffer_index = (int32_t)audio_buffer_read_index;
+        buffer_index = (buffer_index+1)&AUDIO_BUFFER_MASK;
+        int32_t next_sample = (int32_t)((uint8_t *)audio_buffer_ptr)[buffer_index];
+        if (double_pin) {
+            // Convert 0 to 255 to -256 to +254
+            next_sample = next_sample*2-256;
+        }
+       // Sample is set to 7/4 times the input to scale to output interval of 512 cycles.
+        // Actually mutiplied by 28 to account for divide by 16 when generating pulses.
+        next_sample = next_sample*28+8;
+        audio_buffer_read_index = buffer_index;
+        delta = (next_sample-next_value)>>2;
+        if ((buffer_index&(AUDIO_CHUNK_SIZE-1)) == 0 && fetcher_ready) {
+            fetcher_ready = false;
+            set_low_priority_callback(audio_data_fetcher_no_gc, AUDIO_CALLBACK_ID);
+        }
+    }
+    sample = !sample;
+    /* Need to be triggered twice per sample. */
+    return TICK_PER_SAMPLE/2;
+}
+
+static void audio_set_pins(mp_obj_t pin0_obj, mp_obj_t pin1_obj) {
+    PinName p0 = microbit_obj_get_pin_name(pin0_obj);
+    if (pin1_obj == mp_const_none) {
+        init_pin(p0);
+    } else {
+        PinName p1 = microbit_obj_get_pin_name(pin1_obj);
+        init_pins(p0, p1);
+    }
+}
+
+static int32_t pin_read_digital(mp_obj_t pin) {
+    PinName p = microbit_obj_get_pin_name(pin);
+    nrf_gpio_cfg_input(p, NRF_GPIO_PIN_NOPULL);
+    // Allow 1µs to settle.
+    nrf_delay_us(1);
+    return nrf_gpio_pin_read(p);
+}
+
+static const mp_obj_t big_pins[3] = { (mp_obj_t)&microbit_p0_obj, (mp_obj_t)&microbit_p1_obj, (mp_obj_t)&microbit_p2_obj };
+
+static void audio_auto_set_pins(void) {
```

Comment:
I don't think this functions should search all pins.  What if you have a device connected to pin1 which shouldn't have a voltage source applied (eg you connected pin1 to pin2 with a wire and have pin2 in output mode, pin1 in input mode)?  Or you have something connected to pin1/pin2 which is set up and running (eg servo on pin1, or a i2c/uart/spi on pin1/pin2)?

I think it should work as follows: if both pin and return_pin are None then it uses the default, which is to have pin=pin0, and then return_pin either None, pin1 or pin2.  This function can then safely output on pin0 (since the user should at least have 1 of the wires of the speaker connected to pin0) and input on the other pins to see if they change.  Although there is still the issue that the other pins (pin1/pin2) need to already be in input mode for it to work.  You can't change their mode because they may be doing something else (eg servo PWM).